ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,428 @@
+sub-workspaces. Now, without KCP, we need a new design.
+
+The [Environment] CR serves two purposes:
+
+- First, it represents a request from the user for HAS and the GitOps service to **recognize a new destination** for deployments. A new [Environment], when bound to an Application and a Snapshot, causes HAS to write a new directory matching the [Environment] in the gitops repo, which in turn causes Argo to deploy content from that directory somewhere.
+- Second, it represents a request from the user to **provision a new deployment target**. Back in the KCP design, a new [Environment] caused the environment controller(s) to create a new sub-workspace, initialize it, and report back a serviceaccount kubeconfig to be used by Argo to administer it.
+
+Some use cases to consider for [Environments]:
+
+1.  As a part of the StoneSoup workspace initialization process, the user should find that both a **dev and stage environment** with corresponding deployment targets are ready for them ([STONE-180](https://issues.redhat.com/browse/STONE-180)). In our post-KCP architecture, these will be backed by **namespaces** on a devsandbox member cluster.
+2.  The user will want to manually create **additional** [Environments] (for example, a prod environment). The user may want to use our compute resources provided in the form of a new **namespace on a devsandbox member cluster** for this ([STONE-183](https://issues.redhat.com/browse/STONE-183)) or they may want to **bring their own cluster** as a target ([STONE-162](https://issues.redhat.com/browse/STONE-162)).
+3.  The integration-service expects to be able to create **ephemeral** [Environments] for automated testing purposes ([STONE-114](https://issues.redhat.com/browse/STONE-114)). For our short-term goals, the automated testing use case requires the same kind of compute as for the dev and stage [Environments] (devsandbox member cluster namespaces), but will expand to include other kinds of deployment targets in the future - like hypershift clusters ([STONE-185](https://issues.redhat.com/browse/STONE-185)).
+
+## Decision
+
+We are going to emulate storage management APIs (see [persistent-volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and [design.md](https://github.com/kubernetes-csi/external-provisioner/blob/master/doc/design.md) for reference).
+
+### CRDs
+
+#### DeploymentTarget (DT)
+
+A deployment target, usually a K8s api endpoint. The credentials for connecting to the target will
+be stored in a secret which will be referenced in the clusterCredentialsSecret field. A [DT] Can be
+created manually by a user, or dynamically using a provisioner.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTarget
+metadata:
+    name: prod-dt
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+    kubernetesCredentials:
+        defaultNamespace: team-a--prod-dtc
```

Comment:
Just out of curiosity, what does this naming convention mean? I assume team-a is just an imaginary org; what's the rest?

Reply:
Yeah, the convention isn't important, in that nothing depends on parsing this. It could be a uuid or something similar. It just needs to be unique.

But, as the doc was coming together, I was assuming that this would be put together from the `<name of the user's appstudio namespace>--<name of the deploymenttargetclaim in the user's appstudio namespace that caused the new deployment namespace to be created>`